src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java

```diff
@@ -7,297 +7,409 @@
+                                                                                    .collect(Collectors.toMap(
+                                                                                        computedContingencyElement -> computedContingencyElement.getElement().getId(),
+                                                                                        computedContingencyElement -> computedContingencyElement
+                                                                                    ));
+
+        ComputedContingencyElement.setGlobalIndexes(contingenciesElements.values(), factorsGroups.size());
 
         // create jacobian matrix either using base network calculated voltages or nominal voltages
         VoltageInitializer voltageInitializer = lfParameters.getVoltageInitMode() == LoadFlowParameters.VoltageInitMode.PREVIOUS_VALUES ? new PreviousValueVoltageInitializer()
                 : new UniformValueVoltageInitializer();
         JacobianMatrix j = createJacobianMatrix(equationSystem, voltageInitializer);
+        // compute pre-contingency sensitivity values
+        DenseMatrix rhs = initRhs(lfNetwork, equationSystem, factorsGroups, contingenciesElements);
+        DenseMatrix states = solveTransposed(rhs, j); // states contains angles for the sensitivity factors, but also for the +1-1 related to contingencies
+
+        // sensitivities without contingency
+        List<SensitivityValue> sensitivityValues = calculateSensitivityValues(lfNetwork, equationSystem, factorsGroups,
+                states, functionReferenceByBranch, Collections.emptyList(), Collections.emptyMap());
 
-        // solve system
-        DenseMatrix states = solveTransposed(rhs, j);
+        Collection<Contingency> straightforwardContingencies = new LinkedList<>();
+        Map<List<Set<ComputedContingencyElement>>, List<Contingency>> losingConnectivityContingency = new HashMap<>();
+
+        lookForConnectivityLoss(lfNetwork, states, contingencies, contingenciesElements, equationSystem, straightforwardContingencies, losingConnectivityContingency);
+
+        Map<String, List<SensitivityValue>> contingenciesValue = new HashMap<>();
+        // compute the contingencies with no loss of connectivity
+        for (Contingency contingency : straightforwardContingencies) {
+            contingenciesValue.put(contingency.getId(), calculateSensitivityValues(lfNetwork, equationSystem, factorsGroups,
+                    states, functionReferenceByBranch, contingency.getElements().stream().map(element -> contingenciesElements.get(element.getId())).collect(Collectors.toList()),
+                    Collections.emptyMap()
+            ));
+        }
+
+        for (Map.Entry<List<Set<ComputedContingencyElement>>, List<Contingency>> contingencyEntry : losingConnectivityContingency.entrySet()) {
+            Map<SensitivityFactor, Double> predefinedResults = new HashMap<>();
```

Comment:
What is the purpose of predefinedResults? Is it for part of the grid where connectivity has been lost where we want to overwritte results with zero?

Reply:
Yes